### PR TITLE
Parse input value not strictly matching time format in DatePicker component

### PR DESCRIPTION
**Pull request by luca-rath**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | no
| BC breaks? | no
| Deprecations? | no

#### What's in this PR?

This PR adds the functionality to the DatePicker component to enter e.g. "9" in a date picker field with format "HH:mm". Previously the DatePicker component called the onChange callback with undefined then, now it tries to parse that string.

### CODE REVIEW

**Comment by luca-rath** [comment]
@trickreich

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/DatePicker/DatePicker.js [hunk 1]
```diff
@@ -102,6 +102,14 @@ export default class DatePicker extends React.Component<Props> {
 
     handleInputBlur = () => {
         if (this.inputChanged && typeof this.value === 'string') {
+            const fieldOptions = this.getFieldOptions(this.props.options);
+            const currentMoment = moment(this.value, this.getPlaceholder(fieldOptions));
```
Comment (trickreich) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/DatePicker/DatePicker.js]: isn't it more a `newMoment`?

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/DatePicker/DatePicker.js [hunk 2]
```diff
@@ -102,6 +102,14 @@ export default class DatePicker extends React.Component<Props> {
...
+
+            if (currentMoment.isValid()) {
+                this.handleChange(currentMoment.toDate());
```
Comment (trickreich) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/DatePicker/DatePicker.js]: Maybe one line:
`this.handleChange(currentMoment.isValid() ? currentMoment.toDate() : undefined);`?

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/DatePicker/DatePicker.js [hunk 3]
```diff
@@ -102,6 +102,14 @@ export default class DatePicker extends React.Component<Props> {
 
     handleInputBlur = () => {
         if (this.inputChanged && typeof this.value === 'string') {
+            const fieldOptions = this.getFieldOptions(this.props.options);
```
Comment (trickreich) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/DatePicker/DatePicker.js]: Maybe we should change `fieldOptions`?
It's more the like component props for the datepicker component, or am I wrong?

**Review by trickreich — CHANGES_REQUESTED** [review]

**Comment by trickreich** [comment]
@danrot For me this is fine.

**Review by trickreich — APPROVED** [review]

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/DatePicker/DatePicker.js [hunk 4]
```diff
@@ -19,7 +19,11 @@ type Props = {|
     /** Listen for changes of this component */
     onChange: (value: ?Date) => void,
     /** Configure the datepicker to your needs, for more information have a look in the README.md */
-    options: Object,
+    options: {
+        [any]: any,
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/DatePicker/DatePicker.js]: Why is that necessary? I thought as long as you don't put the pipes in the object definition it should allow additional properties without this line.

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/DatePicker/DatePicker.js [hunk 5]
```diff
@@ -121,17 +127,31 @@ export default class DatePicker extends React.Component<Props> {
...
+        }
+
+        return timeFormat || '';
+    };
+
+    getFormat = (): string => {
+        return [
+            this.getDateFormat(),
+            this.getTimeFormat(),
+        ].filter((format) => !!format).join(' ');
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/DatePicker/DatePicker.js]: That's a nice way of doing things! :+1:

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/DatePicker/DatePicker.js [hunk 6]
```diff
@@ -121,17 +127,31 @@ export default class DatePicker extends React.Component<Props> {
...
-        const placeholderTime = fieldOptions.timeFormat ? moment.localeData().longDateFormat('LT') : '';
+    getDateFormat = (): string => {
+        let dateFormat = this.props.options.dateFormat;
 
-        if (!fieldOptions.timeFormat) {
-            return placeholderDate;
-        } else if (fieldOptions.dateFormat && fieldOptions.timeFormat) {
-            return placeholderDate + ' ' + placeholderTime;
+        if ((!dateFormat && dateFormat !== false) || dateFormat === true || (!dateFormat && !this.getTimeFormat())) {
+            dateFormat = moment.localeData().longDateFormat('L');
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/DatePicker/DatePicker.js]: Couldn't you just return here? Would allow to make the `dateFormat` above to be `const`. Or is it possible that this `longDateFormat` call returns a falsey value?

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/DatePicker/DatePicker.js [hunk 7]
```diff
@@ -121,17 +127,31 @@ export default class DatePicker extends React.Component<Props> {
...
 
-        return placeholderTime;
+        return dateFormat || '';
+    };
+
+    getTimeFormat = (): string => {
+        let timeFormat = this.props.options.timeFormat;
+
+        if (timeFormat === true) {
+            timeFormat = moment.localeData().longDateFormat('LT');
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/DatePicker/DatePicker.js]: Same as above

**Review by danrot — CHANGES_REQUESTED** [review]
Also, the tests are missing.

**Review by danrot — APPROVED** [review]
